contrib/babelfishpg_tsql/src/pltsql_ruleutils.c

```diff
@@ -369,6 +375,140 @@ tsql_get_constraintdef(PG_FUNCTION_ARGS)
+
+	isfunction = (proc->prokind != PROKIND_PROCEDURE);
+
+	/*
+	 * We always qualify the function name, to ensure the right function gets
+	 * replaced.
+	 */
+	nsp = get_namespace_name(proc->pronamespace);
+	nnsp= get_logical_schema_name(nsp,true);
+	appendStringInfo(&buf, "CREATE %s %s(",
+					 isfunction ? "FUNCTION" : "PROCEDURE",
+					 tsql_quote_qualified_identifier(nnsp, name));
+	
+	/* we will not pfree name because as we can see name = NameStr(proc->proname) 
+         * here we are not allocating extra space for name, we’re just using proc-> proname.
+         * also at the end, we’re releasing proctup (that will free proc->proname).  
+         */
+	pfree(nsp);
+	pfree(nnsp);
+        
+	tmp = SysCacheGetAttr(PROCOID, proctup, Anum_pg_proc_probin, &isnull);
+        number_args=proc->pronargs; 
+        if(isfunction) number_args++;
+       	probin_json_reader(tmp, &typmod_arr, number_args);
+	(void) print_function_arguments(&buf, proctup, false, true, &typmod_arr);
+	appendStringInfoString(&buf, ")");
+	if (isfunction)
+	{
+		appendStringInfoString(&buf, " RETURNS ");
+		print_function_rettype(&buf, proctup, &typmod_arr, number_args);
+	}
+        if(typmod_arr)
+		pfree(typmod_arr);
+ 
+	/* Emit some miscellaneous options on one line */
+	oldlen = buf.len;
+
+        if (proc->proisstrict)
+		appendStringInfoString(&buf, " WITH RETURNS NULL ON NULL INPUT");
+
+	/* And finally the function definition ... */
+	(void) SysCacheGetAttr(PROCOID, proctup, Anum_pg_proc_prosqlbody, &isnull);
+	
+        appendStringInfoString(&buf, " AS ");
+	tmp = SysCacheGetAttr(PROCOID, proctup, Anum_pg_proc_prosrc, &isnull);
+	prosrc = TextDatumGetCString(tmp); 
+	appendStringInfoString(&buf, prosrc);
+
+	ReleaseSysCache(proctup);
+
+        pfree(prosrc);
+	
+	PG_RETURN_TEXT_P(string_to_text(buf.data));
+}
+
+PG_FUNCTION_INFO_V1(tsql_get_returnTypmodValue);
+/*
+ * function that will return the typmod value of return type
+ */ 
+Datum 
+tsql_get_returnTypmodValue(PG_FUNCTION_ARGS){
+        Oid                     funcid = PG_GETARG_OID(0);
+        HeapTuple      proctup;
+        Form_pg_proc proc;
+        bool            isfunction;
+        Datum           tmp;
+        bool       isnull;
+        int* typmod_arr = NULL;
+        int number_args;
+
+        proctup = SearchSysCache1(PROCOID, ObjectIdGetDatum(funcid));
+        if (!HeapTupleIsValid(proctup)) 
+                PG_RETURN_INT32(-1);
+
+        proc = (Form_pg_proc) GETSTRUCT(proctup);   
+
+        isfunction = (proc->prokind != PROKIND_PROCEDURE);
+
+        if (!isfunction)  
+                PG_RETURN_INT32(-1) ;
+
+        tmp = SysCacheGetAttr(PROCOID, proctup, Anum_pg_proc_probin, &isnull);
+        number_args=proc->pronargs;
+        number_args++;  
+
+        probin_json_reader(tmp, &typmod_arr, number_args);
+
+        if (typmod_arr[number_args-1] != -1)
+               typmod_arr[number_args-1] += adjustTypmod(proc->prorettype, typmod_arr[number_args-1]);
```

Comment:
do we need to free typmod array

Reply:
Yes we need to pfree typmod array.

Reply:
where do we free it?

Reply:
not here because we are returning the value from typmod_arr, in tsql_get_functiondef function we need to pfree typmod array which we have done.